sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/AbstractEventSourcedEntityHandler.scala

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.akkaserverless.javasdk.impl.eventsourcedentity
+
+import com.akkaserverless.javasdk.ServiceCallFactory
+import com.akkaserverless.javasdk.eventsourcedentity.CommandContext
+import com.akkaserverless.javasdk.eventsourcedentity.EventContext
+import com.akkaserverless.javasdk.eventsourcedentity.EventSourcedEntityBase
+import com.akkaserverless.javasdk.eventsourcedentity.EventSourcedEntityCreationContext
+import com.akkaserverless.javasdk.impl.EntityExceptions.EntityException
+import com.akkaserverless.javasdk.impl.effect.SecondaryEffectImpl
+import com.akkaserverless.javasdk.impl.eventsourcedentity.AbstractEventSourcedEntityHandler.CommandResult
+import com.akkaserverless.javasdk.impl.eventsourcedentity.EventSourcedEntityEffectImpl.EmitEvents
+import com.akkaserverless.javasdk.impl.eventsourcedentity.EventSourcedEntityEffectImpl.NoPrimaryEffect
+import com.google.protobuf.{Any => JavaPBAny}
+
+import java.util.Optional
+
+object AbstractEventSourcedEntityHandler {
+  case class CommandResult(events: Vector[Any],
+                           secondaryEffect: SecondaryEffectImpl,
+                           snapshot: Option[Any],
+                           endSequenceNumber: Long)
+}
+
+/**
+ * @tparam S the type of the managed state for the entity
+ * Not for manual user extension or interaction
+ */
+abstract class AbstractEventSourcedEntityHandler[S, E <: EventSourcedEntityBase[S]](protected val entity: E) {
+
+  // ugh, remove
+  var snapshotEvery = 0
+
+  private var state: Option[S] = None
+
+  final protected def stateOrEmpty(): S = state match {
+    case None =>
+      val emptyState = entity.emptyState()
+      require(emptyState != null, "Entity empty state is not allowed to be null")
+      state = Some(emptyState)
+      emptyState
+    case Some(state) => state
+  }
+
+  final protected def setState(newState: S): Unit =
+    state = Some(newState)
```

Comment:
Since `setState` is called with the return of an event handler, we can expect users to return `null`. 


```suggestion
  final protected def setState(newState: S): Unit =
    state = Option(newState)
```